allennlp/tests/training/multi_task_trainer_test.py

```diff
@@ -0,0 +1,284 @@
+        """
+        This method should return one epoch worth of batches.
+        """
+        hoppers: Dict[Any, List[Instance]] = defaultdict(list)
+
+        for instance in instances:
+            # Which hopper do we put this instance in?
+            if self.allow_mixed_batches:
+                instance_type = ""
+            else:
+                instance_type = instance.fields[self.type_field_name].metadata  # type: ignore
+
+            hoppers[instance_type].append(instance)
+
+            # If the hopper is full, yield up the batch and clear it.
+            if len(hoppers[instance_type]) >= self._batch_size:
+                yield Batch(hoppers[instance_type])
+                hoppers[instance_type].clear()
+
+        # Deal with leftovers
+        for remaining in hoppers.values():
+            if remaining:
+                yield Batch(remaining)
+
+
+@Model.register("multi-task-test")
+class MyModel(Model):
+    """
+    This model does nothing interesting, but it's designed to
+    operate on heterogeneous instances using shared parameters
+    (well, one shared parameter) like you'd have in multi-task training.
+    """
+    def __init__(self, vocab: Vocabulary) -> None:
+        super().__init__(vocab)
+        self.weight = torch.nn.Parameter(torch.randn(()))
+
+    def forward(self,  # type: ignore
+                dataset: List[str],
+                field_a: torch.Tensor = None,
+                field_b: torch.Tensor = None,
+                field_c: torch.Tensor = None) -> Dict[str, torch.Tensor]:
+        # pylint: disable=arguments-differ
+        loss = torch.tensor(0.0)   # pylint: disable=not-callable
+        if dataset[0] == "a":
+            loss += field_a["tokens"].sum() * self.weight
+        elif dataset[0] == "b":
+            loss -= field_b["tokens"].sum() * self.weight ** 2
+        elif dataset[0] == "c":
+            loss += field_c["tokens"].sum() * self.weight ** 3
+        else:
+            raise ValueError(f"unknown dataset: {dataset[0]}")
+
+        return {"loss": loss}
+
+
+@Trainer.register("multi-task-test")
+class MultiTaskTrainer(Trainer):
+    """
+    A simple trainer that works in our multi-task setup.
+    Really the main thing that makes this task not fit into our
+    existing trainer is the multiple datasets.
+    """
+    def __init__(self,
+                 model: Model,
+                 serialization_dir: str,
+                 iterator: DataIterator,
+                 mingler: DatasetMingler,
+                 optimizer: torch.optim.Optimizer,
+                 datasets: Dict[str, Iterable[Instance]],
+                 num_epochs: int = 10) -> None:
+        super().__init__(model, serialization_dir)
+        self.iterator = iterator
+        self.mingler = mingler
+        self.optimizer = optimizer
+        self.datasets = datasets
+        self.num_epochs = num_epochs
+
+    def train(self) -> Dict:
```

Comment:
Also not something that I'm saying you should actually _do_ for this toy example, but it'd be good to at least think about it, if you haven't.  How do I get the remaining features from `SingleTaskTrainer` in here?  Like, early stopping, tensorboard logging, model checkpointing, etc.  E.g., the actual training step for both cases should be the same - is there something you could pull out of `SingleTaskTrainer` into a utility function that can be called by both of these?